### PR TITLE
Cog downloader

### PR DESCRIPTION
**Pull request by aikocc**

### CODE REVIEW

File: voxelbotutils/cogs/owner_only.py [hunk 1]
```diff
@@ -152,6 +152,38 @@ async def reload(self, ctx:utils.Context, *cog_name:str):
...
+            with open(file_path, "x", encoding="utf-8") as n:
+                n.write(await r.text())
+        except FileExistsError:
+            return await ctx.send(f"The extension you tried to download was already downloaded.")
+        
+        # Load the cog
+        try:
+            self.bot.load_extension(f"cogs.{file_name[:-3]}")
+        except commands.ExtensionNotFound:
+            return await ctx.send(f"Extension could not be found.")
```
Comment (fadedmax) [inline voxelbotutils/cogs/owner_only.py:178]: No need for the message to be a f string on lines 178, 180, 182, 184

File: voxelbotutils/cogs/owner_only.py [hunk 2]
```diff
@@ -152,6 +152,38 @@ async def reload(self, ctx:utils.Context, *cog_name:str):
             await ctx.send('```py\n' + traceback.format_exc() + '```')
             return
         await ctx.send('Cog reloaded.')
+        
+    @commands.command(aliases=['dlcog'], cls=utils.Command)
+    @commands.is_owner()
+    async def downloadcog(self, ctx: utils.Context, url: str):
+        """Download a cog from github"""
+    
+        #Convert github link to a raw link and grab contents
```
Comment (fadedmax) [inline voxelbotutils/cogs/owner_only.py:161]: There should be a space between the # and the comment

File: voxelbotutils/cogs/owner_only.py [hunk 3]
```diff
@@ -152,6 +152,38 @@ async def reload(self, ctx:utils.Context, *cog_name:str):
...
+        raw_url = url.replace("/blob", "").replace("github.com", "raw.githubusercontent.com")
+        r = await self.bot.session.get(raw_url)
```
Comment (fadedmax) [inline voxelbotutils/cogs/owner_only.py:163]: This should have headers with the user agent, probably?

File: voxelbotutils/cogs/owner_only.py [hunk 4]
```diff
@@ -152,6 +152,38 @@ async def reload(self, ctx:utils.Context, *cog_name:str):
             await ctx.send('```py\n' + traceback.format_exc() + '```')
             return
         await ctx.send('Cog reloaded.')
+        
+    @commands.command(aliases=['dlcog'], cls=utils.Command)
+    @commands.is_owner()
+    async def downloadcog(self, ctx: utils.Context, url: str):
```
Comment (fadedmax) [inline voxelbotutils/cogs/owner_only.py:158]: in my opinion i dont see the need for a space between `ctx: and utils.Context`, it doesn't align with the rest of the file.

File: voxelbotutils/cogs/owner_only.py [hunk 5]
```diff
@@ -152,6 +152,38 @@ async def reload(self, ctx:utils.Context, *cog_name:str):
...
+        try:
+            self.bot.load_extension(f"cogs.{file_name[:-3]}")
+        except commands.ExtensionNotFound:
+            return await ctx.send(f"Extension could not be found.")
+        except commands.ExtensionAlreadyLoaded:
+            return await ctx.send(f"The extension you tried to download was already running.")
+        except commands.NoEntryPointError:
+            return await ctx.send(f"No added setup function.")
+        except commands.ExtensionFailed:
+            return await ctx.send(f"Extension failed for some unknown reason.")
```
Comment (fadedmax) [inline voxelbotutils/cogs/owner_only.py:184]: (personal) There should also just be a general error catch here if something else unknown fucked up in the process.

**Review by fadedmax — CHANGES_REQUESTED** [review]

File: voxelbotutils/cogs/owner_only.py (hunk 4 above)
Comment (fadedmax) [reply voxelbotutils/cogs/owner_only.py:158]: so like `async def downloadcog(self, ctx:utils.Context, url:str):`

**Comment by 4Kaylum** [comment]
If the file fails to load, should it be deleted, renamed to have an underscore prefix so it doesn't try to load next startup, or neither?

**Comment by aikocc** [comment]
> If the file fails to load, should it be deleted, renamed to have an underscore prefix so it doesn't try to load next startup, or neither?

I feel like it should be deleted as its not going to allow the file to be overwritten and also if that cog is added on GitHub it will cause errors as is wouldn't have been stashed or whatever is called when GitHub finds a file that is already in the directory.

**Comment by 4Kaylum** [comment]
Fine by me - if cog errors on load extension, delete cog. Ez.

**Comment by fadedmax** [comment]
Yup, warn user that 'ayy ooo bro this cog no work sorryyyy' and then just delete it

**Comment by aikocc** [comment]
> 
> 
> Fine by me - if cog errors on load extension, delete cog. Ez.

Added to command, ready to be merged now.

File: voxelbotutils/cogs/owner_only.py [hunk 6]
```diff
@@ -181,7 +182,8 @@ async def downloadcog(self, ctx: utils.Context, url: str):
         except commands.NoEntryPointError:
             return await ctx.send("No added setup function.")
         except commands.ExtensionFailed:
-            return await ctx.send("Extension failed for some unknown reason.")
+            os.remove(file_path)
+            return await ctx.send("Extension failed for some unknown reason. Extension has been deleted.")
```
Comment (fadedmax) [inline voxelbotutils/cogs/owner_only.py:186]: There should be a comma there, not a period.

Comment (4Kaylum) [reply voxelbotutils/cogs/owner_only.py:186]: I don't think that's necessary.

**Comment by aikocc** [comment]
Added:
- Command will delete file on all errors when loading the extension.

**Pull request merged**